crates/sequencer/src/db_provider/mod.rs

```diff
@@ -173,33 +189,123 @@ impl BlockBodyIndicesProvider for DbProvider {
 }
 
 impl BlockReaderIdExt for DbProvider {
-    fn block_by_id(&self, _id: BlockId) -> ProviderResult<Option<Self::Block>> {
-        unimplemented!("block_by_id")
+    fn block_by_id(&self, id: BlockId) -> ProviderResult<Option<Self::Block>> {
+        let block_number = match id {
+            BlockId::Hash(hash) => {
+                let mut working_set = WorkingSet::new(self.storage.clone());
+                match self.evm.get_block_by_hash(
+                    hash.block_hash,
+                    None,
+                    &mut working_set,
+                    &self.ledger_db,
+                ) {
+                    Ok(Some(block)) => block.header.number,
+                    _ => return Ok(None),
+                }
+            }
+            BlockId::Number(num_or_tag) => match num_or_tag {
+                BlockNumberOrTag::Number(n) => n,
+                BlockNumberOrTag::Latest => self.best_block_number()?,
+                BlockNumberOrTag::Finalized => match self.finalized_block_number()? {
+                    Some(n) => n,
+                    None => return Ok(None),
+                },
+                BlockNumberOrTag::Safe => match self.finalized_block_number()? {
+                    Some(n) => n,
+                    None => return Ok(None),
+                },
```

Comment:
```suggestion
                BlockNumberOrTag::Finalized | BlockNumberOrTag::Safe => match self.finalized_block_number()? {
                    Some(n) => n,
                    None => return Ok(None),
                },
```
Makes it explicit that we use the same logic for Finalized and Safe